### PR TITLE
ZKUI-53 & ZKUI-54: Add legal hold in the object properties

### PR DESCRIPTION
**Pull request by ChengYanJin**
ZKUI-53: Add balance-scale in the object list
ZKUI-54: Add legal hold toggle in the object properties

image
image

### CODE REVIEW

File: src/react/actions/s3object.js [hunk 1]
```diff
@@ -533,3 +543,23 @@ export function putObjectTagging(
...
+  isLegalHold: boolean,
+  prefixWithSlash: string,
+): ThunkStatePromisedAction {
+  return (dispatch, getState) => {
+    const { zenkoClient } = getClients(getState());
+    dispatch(networkStart('Getting object legal hold'));
+    return zenkoClient
+      .putObjectLegalHold(bucketName, objectKey, versionId, isLegalHold)
+      .then(() => dispatch(getObjectMetadata(bucketName, objectKey, versionId)))
+      .then(() => dispatch(listObjects(bucketName, prefixWithSlash)))
```
Comment (JBWatenbergScality) [inline src/react/actions/s3object.js]: Maybe we could just dispatch the action that select the object here temporarily while the url rework is not done ?

File: src/react/databrowser/objects/details/Properties.jsx [hunk 2]
```diff
@@ -17,6 +19,10 @@ const TruncatedValue = styled(T.Value)`
 `;
 
 function Properties({ objectMetadata }: Props) {
+  const dispatch = useDispatch();
+  const prefixWithSlash = usePrefixWithSlash();
+  const isLegalHoldEnabled = objectMetadata.legalHoldStatus === 'ON';
```
Comment (JBWatenbergScality) [inline src/react/databrowser/objects/details/Properties.jsx]: I'm wondering wether we should not keep this ON logic on the client side and everywhere else use a boolean.
In src/js/S3Client.js we could change :
```
getObjectLegalHold(bucketName, objectKey, versionId) {
    const params = {
      Bucket: bucketName,
      Key: objectKey,
      VersionId: versionId,
    };

    return new Promise((resolve, reject) => {
      this.client.getObjectLegalHold(params, (error, data) => {
        if (error) {
          if (
            error.code === 'NoSuchObjectLockConfiguration' ||
            error.code === 'InvalidRequest'
          ) {
            return resolve('OFF');
          }
          return reject(error);
        }
        if (data.Status) {
          return resolve(data.Status);
        }
        return resolve('OFF');
      });
    });
  }
```

To 
```
getObjectLegalHold(bucketName, objectKey, versionId) {
    const params = {
      Bucket: bucketName,
      Key: objectKey,
      VersionId: versionId,
    };

    return new Promise((resolve, reject) => {
      this.client.getObjectLegalHold(params, (error, data) => {
        if (error) {
          if (
            error.code === 'NoSuchObjectLockConfiguration' ||
            error.code === 'InvalidRequest'
          ) {
            return resolve(false);
          }
          return reject(error);
        }
        if (data.Status) {
          return resolve(data.Status === 'ON');
        }
        return resolve(false);
      });
    });
  }
```

Then everywhere else we wouldn't have to take care of the boolean translation of the value

File: src/react/databrowser/objects/details/Properties.jsx [hunk 3]
```diff
@@ -61,14 +67,53 @@ function Properties({ objectMetadata }: Props) {
...
             <T.Key> Lock </T.Key>
             <T.Value>
-              {objectMetadata.lockStatus === 'LOCKED' && <>Locked ({objectMetadata.objectRetention.mode.toLowerCase()})<br />until {objectMetadata.objectRetention.retainUntilDate}</>}
-              {objectMetadata.lockStatus === 'RELEASED' && `Released - since ${objectMetadata.objectRetention.retainUntilDate}`}
+              {objectMetadata.lockStatus === 'LOCKED' && (
+                <>
+                  Locked ({objectMetadata.objectRetention.mode.toLowerCase()})
+                  <br />
+                  until {objectMetadata.objectRetention.retainUntilDate}
+                </>
```
Comment (JBWatenbergScality) [inline src/react/databrowser/objects/details/Properties.jsx]: I think it is missing the locked icon here for the discoverability of the icon meaning (which was the starting point of this refactoring

File: src/react/databrowser/objects/details/Properties.jsx [hunk 4]
```diff
@@ -61,14 +67,53 @@ function Properties({ objectMetadata }: Props) {
...
+              )}
+              {objectMetadata.lockStatus === 'RELEASED' &&
+                `Released - since ${objectMetadata.objectRetention.retainUntilDate}`}
```
Comment (JBWatenbergScality) [inline src/react/databrowser/objects/details/Properties.jsx]: I think the unlocked icon is also missing here

File: src/react/databrowser/objects/details/Properties.jsx [hunk 5]
```diff
@@ -61,14 +67,53 @@ function Properties({ objectMetadata }: Props) {
...
+                        !isLegalHoldEnabled,
+                        prefixWithSlash,
+                      ),
+                    )
+                  }
+                />
+                {isLegalHoldEnabled && (
+                  <i
+                    className="fas fa-balance-scale"
+                    style={{ paddingLeft: '5px' }}
```
Comment (JBWatenbergScality) [inline src/react/databrowser/objects/details/Properties.jsx]: shouldn't we use 0.5rem instead ? Maybe zenko is not yet ready for the rem unit conversion, I'm just wonering

File: src/react/databrowser/objects/details/Properties.jsx [hunk 6]
```diff
@@ -1,5 +1,5 @@
 // @noflow
-import Table, * as T from '../../../ui-elements/TableKeyValue';
+import Table, * as T from '../../../ui-elements/TableKeyValue2';
```
Comment (JBWatenbergScality) [inline src/react/databrowser/objects/details/Properties.jsx:2]: Is TableKeyValue still used after this change ? If not I think we could drop it and rename the TableKeyValue2 to TableKeyValue 🚀

File: src/react/databrowser/objects/details/Properties.jsx (hunk 3 above)
Comment (JBWatenbergScality) [reply src/react/databrowser/objects/details/Properties.jsx]: I've seen it afterwards in commit https://github.com/scality/zenko-ui/pull/214/commits/540df519b27ffa9a9e66cf2d34a0a85fdb9afd9b

File: src/react/databrowser/objects/details/Properties.jsx (hunk 4 above)
Comment (JBWatenbergScality) [reply src/react/databrowser/objects/details/Properties.jsx]: I've seen it afterwards in commit https://github.com/scality/zenko-ui/pull/214/commits/540df519b27ffa9a9e66cf2d34a0a85fdb9afd9b

File: src/react/databrowser/objects/details/Properties.jsx (hunk 6 above)
Comment (ChengYanJin) [reply src/react/databrowser/objects/details/Properties.jsx:2]: Sadly this TableKeyValue is still used by many other components :(

File: src/react/actions/s3object.js (hunk 1 above)
Comment (ChengYanJin) [reply src/react/actions/s3object.js]: that's right!!! I will just reverse L.559 and L.560.  🤩  
and problem solved!

File: src/react/databrowser/objects/details/Properties.jsx (hunk 5 above)
Comment (ChengYanJin) [reply src/react/databrowser/objects/details/Properties.jsx]: I actually changed in the later commit with `spacing`
```
const Icon = styled.i`
  margin-left: ${spacing.sp4};
`;
```

File: src/react/databrowser/objects/details/Properties.jsx (hunk 2 above)
Comment (ChengYanJin) [reply src/react/databrowser/objects/details/Properties.jsx]: Yeah. It makes sense to store the boolean `isLegalHoldEnabled`

File: src/react/databrowser/objects/details/Properties.jsx (hunk 6 above)
Comment (JBWatenbergScality) [reply src/react/databrowser/objects/details/Properties.jsx:2]: Would be for next time then !

**Review by JBWatenbergScality — APPROVED** [review]

File: src/react/databrowser/objects/details/Properties.jsx [hunk 7]
```diff
@@ -1,74 +1,141 @@
...
   objectMetadata: ObjectMetadata,
 };
 
-const TruncatedValue = styled(T.Value)`
-  max-width: 300px;
+const TruncatedValue = styled.div`
+  max-width: 18rem;
+`;
+
+const Icon = styled.i`
```
Comment (nicolas2bert) [inline src/react/databrowser/objects/details/Properties.jsx:23]: NIT: I know ui-elements/Table as an `Icon` component you might want to reuse. Or you can introduce one in `TableKeyValue2`

**Review by nicolas2bert — APPROVED** [review]

**Pull request merged**